torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -328,10 +344,11 @@ auto handle_torch_function_indexing(PyObject* self, PyObject* index, PyObject* v
  *
  */
 
-void append_overloaded_arg(std::vector<py::handle>* overloaded_args, PyObject* obj) {
+static void append_overloaded_arg(std::vector<py::handle>* overloaded_args, PyObject* obj, bool obj_is_type) {
   bool class_not_seen_yet = true;
+  PyObject* obj_type = obj_is_type ? obj : (PyObject*)Py_TYPE(obj);
   for (auto &arg : *overloaded_args) {
-    if (Py_TYPE(obj) == Py_TYPE(arg.ptr())) {
+    if (obj_type == get_type_of_overloaded_arg(arg.ptr())) {
```

Comment:
Blast, needing to check if it's a type or not really sucks. But now I'm wondering... could we possibly make overloaded_args actually be overloaded_types? (I guess the only thing that would break are legacy cases where people `__torch_function__` non-classmethod)

Reply:
Is the recommendation for `__torch_function__` now to use it as a classmethod? (and yes, the legacy cases would break). I'm down to change it to make it overloaded_types (in an earlier version of this PR it was actually  `overloaded_types` instead) and keep alive a legacy path for the legacy case

Reply:
yes. But the method spelling does work and I suspect some people would get broken if we switched it.

Reply:
@ezyang would you be fine if we had two separate paths, one for "overloaded args" and one for "overloaded types"? More concretely, that would look like the following:
- handle_torch_function_no_python_arg_parser uses overloaded_args
- (new!) handle_torch_dispatch_no_python_arg_parser uses overloaded_types
- append_overloaded_arg handles overloaded args
- (new!) append_overloaded_type handles overloaded types

This would involve some code duplication between handle_torch_function_no_python_arg_parser and handle_torch_dispatch_no_python_arg_parser as well as between append_overloaded_arg and append_overloaded_type.

I think this approach is reasonable if we are planning on deprecating the legacy non-classmethod torch_function usages -- when we do that then there won't be code duplication. (is this the plan?)

There's a third alternative that looks something like "we turn overloaded_args from a `vector<py::handle>` to `vector<PyObjOrType>`" where `PyObjOrType` is a struct that holds a `py::handle` and a bool. This solves a lot of issues with performing extra PyType_checks and would be better if we're planning on keeping around the legacy torch_function path around for a while.

Reply:
I guess I'm missing information about how many people are incorrectly doing `__torch_function__` as a non-classmethod. If there aren't many I think we can expire this code very quickly, and then I agree with your analysis. I'm not very keen on doing more complicated things to remove PyType_ checks, would rather just regress performance instead.

Reply:
I did a quick github search. I found 40 usages of `__torch_function__`, of which [77% are using the legacy style](https://docs.google.com/spreadsheets/d/1M3QIrPPugCmvQWaYH2k8dUKWtcZ5j-_buNHyAToewFU/edit?usp=sharing). So I don't think we can deprecate the non-classmethod `__torch_function__` soon, we probably want to do a full deprecation cycle and help migrate some of the more popular users.

Given that, I'm planning on sticking with the strategy that is in this PR that regresses performance